### PR TITLE
Bootstrap make reproducibly

### PR DESCRIPTION
**Pull request by fmeum**
Uses the Bazel C/C++ toolchain to bootstrap make and ensure that the
resulting binary contains no absolute and thus non-hermetic paths.

Building make reproducibly helps with remote caching and removes the
dependency on a C compiler installed on the host.

### CODE REVIEW

File: foreign_cc/built_tools/make_build.bzl
Comment (jheaff1) [inline foreign_cc/built_tools/make_build.bzl:91]: To resolve the build failure on MacOS, you may have to unset the `AR` variable, as is done in a few of the third party examples in this repo

File: foreign_cc/built_tools/make_build.bzl
Comment (jheaff1) [inline foreign_cc/built_tools/make_build.bzl:33]: Did you intentionally remove the `—with-guile=no` argument?

Comment (fmeum) [reply foreign_cc/built_tools/make_build.bzl:33]: I didn't, thanks for catching it.

File: foreign_cc/built_tools/make_build.bzl
Comment (jheaff1) [reply foreign_cc/built_tools/make_build.bzl]: To be honest I don’t really know why `AR` has to be unset for MacOS. Perhaps @UebelAndre or @jsharpe could offer some insight?

Comment (jsharpe) [reply foreign_cc/built_tools/make_build.bzl]: I'm not entirely sure either but suspect its an incompatibility between setting ARFLAGS to be the output from the cc toolchain's `cxx_linker_static`  probably because the native link static rule adds additional flags somewhere else but I don't know the mechanics of the toolchain on macOS well enough to be able to say that with complete certainty.

@fmeum would you mind separating this PR into 2 to make the changes to `AR` on macOS separately so that it can be reverted independently if it causes any breakage that the examples in this repo don't show?

Comment (jsharpe) [reply foreign_cc/built_tools/make_build.bzl]: Oh wait - you already did - my bad; I should read the full PR before commenting!

File: foreign_cc/built_tools/make_build.bzl
Comment (jsharpe) [inline foreign_cc/built_tools/make_build.bzl:85]: Can this be replaced by a call to `get_flags_info` from `foreign_cc/private/cc_toolchain_util.bzl`; there are a few standard features that need to be disabled in the feature configuration that this code already does.
Also we should probably also be grabbing any environment variables from the toolchain too - again `get_env_vars` should provide this list.

File: toolchains/make-reproducible-bootstrap.patch
Comment (jsharpe) [inline toolchains/make-reproducible-bootstrap.patch:12]: I appreciate this actually makes this reproducible but does this actually work to allow make to correctly find its locale directory? The configure_make rules run in `$BUILDTMP` which is below the bazel execroot and so this relative path is not going to be correct from that directory. 
I don't have a good solution for this though that is any better than what you've already done here; I suspect we don't really care that the make tool can't find its translations but we should at least try to ensure that the tool does work (Which I appreciate even in the current configuration isn't the case due to sandbox paths, but it would currently at least currently work in a non-sandboxed local execution environment).

**Review by jsharpe — CHANGES_REQUESTED** [review]
@jheaff1 Thanks for the initial review comments on this.
@fmeum Thanks for this work; making the make build reproducible is something that wasn't even on my radar of tasks to do!

Can I request that the toolchain configuration reaches out to the functions already defined for this purpose? That way we aren't maintaining 2 copies of essentially the same logic. Also we need to propagate across any environment variables from the toolchain configuration as sometimes compiler wrappers may set these values in the toolchain configuration.

File: foreign_cc/built_tools/make_build.bzl
Comment (fmeum) [reply foreign_cc/built_tools/make_build.bzl]: Done. This required small changes to `cc_toolchain_utils.bzl` which I believe to be generally useful. I added them as a separate commit.

File: foreign_cc/built_tools/make_build.bzl
Comment (fmeum) [reply foreign_cc/built_tools/make_build.bzl]: Sort of - I still kept the change to `AR` in this PR for now, but would wait until I got #819 to work and merged. Afterwards, I would be able to delete the logic from this PR.

File: toolchains/make-reproducible-bootstrap.patch
Comment (fmeum) [reply toolchains/make-reproducible-bootstrap.patch:12]: From what I can tell, these directories default to `/lib`, `/include` and `/share/locale`, so they are not `make`-specific. Based on a cursory look into the code, these variables are used for the following purposes:
* `LIBDIR` is only used in `remake.c`, apparently to find out whether a certain system library has changed since the last `make` invocation. This should not be needed in the context of `rules_foreign_cc`.
* `INCLUDEDIR` is the default search path for `Makefile`s to include, which can also be specified with the `-I` option to `make`. Picking up `Makefile`s from the host in unsandboxed builds sounds really risky, so maybe this should always be set to something like `/does/not/exist`.
* `LOCALEDIR` is used to make `make` translatable.

While I do favor including this patch as it helps reproducibility and these directories are already non-existent in current sandboxed builds, I can drop it if you prefer.

**Review by jsharpe — APPROVED** [review]

**Pull request merged**
